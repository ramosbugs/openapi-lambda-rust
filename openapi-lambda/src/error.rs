use crate::{HeaderName, HttpResponse, StatusCode};

use aws_lambda_events::encodings::Body;
// Until std::error::Backtrace is fully stabilized, we can't embed a type named `Backtrace` within
// a thiserror::Error (see https://github.com/dtolnay/thiserror/issues/204).
use backtrace::Backtrace as _Backtrace;
use headers::{ContentType, Header};
use itertools::Itertools;
use log::error;
use thiserror::Error;

use std::borrow::Cow;
use std::string::FromUtf8Error;

/// Error that occurred while processing an AWS Lambda event.
#[non_exhaustive]
#[derive(Debug, Error)]
pub enum EventError {
  /// Failed to prepare HTTP response.
  #[error("failed to prepare HTTP response")]
  HttpResponse(#[source] Box<http::Error>, _Backtrace),
  /// Invalid base64 encoding for request body.
  // The base64 encoding comes from AWS, so this is actually an internal error.
  #[error("invalid base64 encoding for request body")]
  InvalidBodyBase64(#[source] Box<base64::DecodeError>, _Backtrace),
  /// Failed to JSON deserialize request body.
  #[error("failed to JSON deserialize request body")]
  InvalidBodyJson(
    #[source] Box<serde_path_to_error::Error<serde_json::Error>>,
    _Backtrace,
  ),
  /// Invalid UTF-8 encoding for request body.
  #[error("invalid UTF-8 encoding for request body")]
  InvalidBodyUtf8(#[source] Box<FromUtf8Error>, _Backtrace),
  /// Invalid UTF-8 encoding for request header.
  #[error("invalid UTF-8 encoding for request header `{0}`")]
  InvalidHeaderUtf8(
    HeaderName,
    // We don't use `http::header::ToStrError` here since aws_lambda_events uses a different version
    // of `http`.
    #[source] Box<dyn std::error::Error + Send + Sync + 'static>,
    _Backtrace,
  ),
  /// Failed to parse request path parameter.
  #[error("failed to parse request path parameter `{param_name}`")]
  InvalidRequestPathParam {
    /// Name of the parameter that failed to parse.
    param_name: Cow<'static, str>,
    /// Underlying error that occurred while parsing the param.
    #[source]
    source: Option<Box<dyn std::error::Error + Send + Sync + 'static>>,
    /// Stack trace indicating where the error occurred.
    backtrace: _Backtrace,
  },
  /// Failed to parse request query param.
  #[error("failed to parse request query param `{param_name}`")]
  InvalidRequestQueryParam {
    /// Name of the parameter that failed to parse.
    param_name: Cow<'static, str>,
    /// Underlying error that occurred while parsing the param.
    #[source]
    source: Option<Box<dyn std::error::Error + Send + Sync + 'static>>,
    /// Stack trace indicating where the error occurred.
    backtrace: _Backtrace,
  },
  /// Missing required request body.
  #[error("missing required request body")]
  MissingRequestBody(_Backtrace),
  /// Missing required request header.
  #[error("missing required request header `{0}`")]
  MissingRequestHeader(Cow<'static, str>, _Backtrace),
  /// Missing required request param.
  #[error("missing required request param `{0}`")]
  MissingRequestParam(Cow<'static, str>, _Backtrace),
  /// Request handler panicked.
  #[error("request handler panicked: {0}")]
  Panic(String, _Backtrace),
  /// Failed to serialize response body to JSON.
  #[error("failed to serialize {type_name} response to JSON")]
  ToJsonResponse {
    /// Name of the response body type that failed to serialize.
    type_name: Cow<'static, str>,
    /// Underlying error that occurred while serializing the response body.
    #[source]
    source: Box<serde_path_to_error::Error<serde_json::Error>>,
    /// Stack trace indicating where the error occurred.
    backtrace: _Backtrace,
  },
  /// Unexpected request Content-Type.
  #[error("unexpected Content-Type `{0}`")]
  UnexpectedContentType(String, _Backtrace),
  /// Unexpected operation ID.
  #[error("unexpected operation ID: {0}")]
  UnexpectedOperationId(String, _Backtrace),
}

impl EventError {
  /// Return the backtrace associated with the error, if known.
  pub fn backtrace(&self) -> Option<&_Backtrace> {
    match self {
      EventError::HttpResponse(_, backtrace)
      | EventError::InvalidBodyBase64(_, backtrace)
      | EventError::InvalidBodyJson(_, backtrace)
      | EventError::InvalidBodyUtf8(_, backtrace)
      | EventError::InvalidHeaderUtf8(_, _, backtrace)
      | EventError::InvalidRequestPathParam { backtrace, .. }
      | EventError::InvalidRequestQueryParam { backtrace, .. }
      | EventError::MissingRequestBody(backtrace)
      | EventError::MissingRequestHeader(_, backtrace)
      | EventError::MissingRequestParam(_, backtrace)
      | EventError::Panic(_, backtrace)
      | EventError::ToJsonResponse { backtrace, .. }
      | EventError::UnexpectedContentType(_, backtrace)
      | EventError::UnexpectedOperationId(_, backtrace) => Some(backtrace),
    }
  }

  /// Return the name of the error variant (e.g., `InvalidBodyBase64`).
  pub fn name(&self) -> &str {
    match self {
      EventError::HttpResponse(_, _) => "HttpResponse",
      EventError::InvalidBodyBase64(_, _) => "InvalidBodyBase64",
      EventError::InvalidBodyJson(_, _) => "InvalidBodyJson",
      EventError::InvalidBodyUtf8(_, _) => "InvalidBodyUtf8",
      EventError::InvalidHeaderUtf8(_, _, _) => "InvalidHeaderUtf8",
      EventError::InvalidRequestPathParam { .. } => "InvalidRequestPathParam",
      EventError::InvalidRequestQueryParam { .. } => "InvalidRequestQueryParam",
      EventError::MissingRequestBody(_) => "MissingRequestBody",
      EventError::MissingRequestHeader(_, _) => "MissingRequestHeader",
      EventError::MissingRequestParam(_, _) => "MissingRequestParam",
      EventError::Panic(_, _) => "Panic",
      EventError::ToJsonResponse { .. } => "ToJsonResponse",
      EventError::UnexpectedContentType(_, _) => "UnexpectedContentType",
      EventError::UnexpectedOperationId(_, _) => "UnexpectedOperationId",
    }
  }
}

impl From<EventError> for HttpResponse {
  /// Build a client-facing [`HttpResponse`] appropriate for the error that occurred.
  ///
  /// This function will set the appropriate HTTP status code (400 or 500) depending on whether the
  /// error is internal (500) or caused by the client (400). For client errors, the
  /// response body contains a human-readable description of the error and the `Content-Type`
  /// response header is set to `text/plain`. For internal errors, no response body is returned to
  /// the client.
  fn from(err: EventError) -> HttpResponse {
    let (status_code, body) = match err {
      // 400
      EventError::InvalidBodyJson(err, _) => (
        StatusCode::BAD_REQUEST,
        // We expose parse errors to the client to provide better 400 Bad Request diagnostics.
        Some(if err.path().iter().next().is_none() {
          format!("Invalid request body: {}", err.inner())
        } else {
          format!(
            "Invalid request body (path: `{}`): {}",
            err.path(),
            err.inner()
          )
        }),
      ),
      EventError::InvalidBodyUtf8(_, _) => (
        StatusCode::BAD_REQUEST,
        Some("Request body must be UTF-8 encoded".to_string()),
      ),
      EventError::InvalidHeaderUtf8(header_name, _, _) => (
        StatusCode::BAD_REQUEST,
        Some(format!(
          "Invalid value for header `{header_name}`: must be UTF-8 encoded"
        )),
      ),
      EventError::InvalidRequestPathParam { param_name, .. } => (
        StatusCode::BAD_REQUEST,
        Some(format!("Invalid `{param_name}` request path parameter")),
      ),
      EventError::InvalidRequestQueryParam { param_name, .. } => (
        StatusCode::BAD_REQUEST,
        Some(format!("Invalid `{param_name}` query parameter")),
      ),
      EventError::MissingRequestBody(_) => (
        StatusCode::BAD_REQUEST,
        Some("Missing request body".to_string()),
      ),
      EventError::MissingRequestHeader(header_name, _) => (
        StatusCode::BAD_REQUEST,
        Some(format!("Missing request header `{header_name}`")),
      ),
      EventError::MissingRequestParam(param_name, _) => (
        StatusCode::BAD_REQUEST,
        Some(format!("Missing required parameter `{param_name}`")),
      ),
      EventError::UnexpectedContentType(content_type, _) => (
        StatusCode::BAD_REQUEST,
        Some(format!("Unexpected content type `{content_type}`")),
      ),
      // 500
      EventError::HttpResponse(_, _)
      | EventError::InvalidBodyBase64(_, _)
      | EventError::Panic(_, _)
      | EventError::ToJsonResponse { .. }
      | EventError::UnexpectedOperationId(_, _) => (StatusCode::INTERNAL_SERVER_ERROR, None),
    };

    let mut response = if let Some(body_str) = body {
      error!("Responding with error status {status_code}: {body_str}");

      let mut response = HttpResponse::new(Body::Text(body_str));
      response.headers_mut().insert(
        ContentType::name().to_owned(),
        ContentType::text()
          .to_string()
          .try_into()
          .expect("MIME type should be a valid header"),
      );

      response
    } else {
      error!("Responding with error status {status_code}");

      HttpResponse::new(Body::Empty)
    };

    *response.status_mut() = status_code;

    response
  }
}

/// Helper function for formatting an error as a string containing a human-readable chain of causes.
///
/// This function will walk over the chain of causes returned by
/// [`Error::source`](std::error::Error::source) and append each underlying error (using the
/// [`Display`](std::fmt::Display) trait).
///
/// # Arguments
///
/// * `err` - Error to format.
/// * `name` - Optional name of the error type/variant (e.g., `EventError::InvalidBodyJson`).
/// * `backtrace` - Optional [`Backtrace`](backtrace::Backtrace) indicating where the top-level
///   error occurred.
pub fn format_error(
  err: &(dyn std::error::Error),
  name: Option<&str>,
  backtrace: Option<&_Backtrace>,
) -> String {
  let err_line = name
    .map(|n| format!("{}: {}", n, err))
    .unwrap_or_else(|| err.to_string());

  let top_error = if let Some(bt) = backtrace {
    format!("{err_line}\n  stack trace:\n{}", format_backtrace(bt, 4))
  } else {
    err_line
  };

  let cause_str = ErrorCauseIterator(err.source())
    .map(|cause| format!("  caused by: {cause}"))
    .join("\n");

  if !cause_str.is_empty() {
    format!("{top_error}\n{cause_str}")
  } else {
    top_error
  }
}

struct ErrorCauseIterator<'a>(Option<&'a (dyn std::error::Error + 'static)>);

impl<'a> Iterator for ErrorCauseIterator<'a> {
  type Item = &'a (dyn std::error::Error + 'static);

  fn next(&mut self) -> Option<Self::Item> {
    let current = self.0;
    self.0 = current.and_then(|err| err.source());
    current
  }
}

fn format_backtrace(backtrace: &_Backtrace, indent: usize) -> String {
  let indent_str = " ".repeat(indent);
  format!("{backtrace:?}")
    .lines()
    .join(&format!("{indent_str}\n"))
}
