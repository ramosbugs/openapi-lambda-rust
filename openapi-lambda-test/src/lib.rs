#![allow(unused_variables)]

include!(concat!(env!("OUT_DIR"), "/out.rs"));

mod types {
  use std::num::ParseIntError;
  use std::str::FromStr;

  #[derive(Clone, Copy, Debug)]
  pub struct BarId(i64);

  impl FromStr for BarId {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
      i64::from_str(s).map(Self)
    }
  }
}

// Make sure the auto-generated handler templates compile.
#[allow(dead_code)]
#[path = "../.openapi-lambda/foo_handler.rs"]
pub mod foo_handler;

#[path = "../.openapi-lambda/bar_handler.rs"]
#[allow(dead_code)]
pub mod bar_handler;

// TO UPDATE THE OUTPUT SNAPSHOTS BELOW, RUN:
//   cargo insta test --review
// This requires having previously run `cargo install cargo-insta`.
// See: https://insta.rs/docs/quickstart/
#[cfg(test)]
mod tests {
  use insta::{assert_display_snapshot, assert_yaml_snapshot};
  use openapi_lambda::error::format_error;
  use openapiv3::OpenAPI;
  use proc_macro2::TokenStream;

  use std::fs::File;
  use std::path::Path;

  #[test]
  fn test_openapi_apigw() {
    let openapi_apigw_path = Path::new(".openapi-lambda/openapi-apigw.yaml");
    let openapi_apigw_contents =
      serde_path_to_error::deserialize::<_, OpenAPI>(serde_yaml::Deserializer::from_reader(
        &File::open(openapi_apigw_path)
          .unwrap_or_else(|err| panic!("failed to read {}: {err}", openapi_apigw_path.display())),
      ))
      .unwrap_or_else(|err| {
        panic!(
          "failed to parse {}: {}",
          openapi_apigw_path.display(),
          format_error(&err, None, None)
        )
      });
    assert_yaml_snapshot!("openapi-apigw.yaml", openapi_apigw_contents);
  }

  #[test]
  fn test_out_rs() {
    let out_rs_path = Path::new(concat!(env!("OUT_DIR"), "/out.rs"));
    let out_rs_contents = std::fs::read_to_string(out_rs_path)
      .unwrap_or_else(|err| panic!("failed to read {}: {err}", out_rs_path.display()));
    out_rs_contents
      .parse::<TokenStream>()
      .unwrap_or_else(|err| {
        panic!(
          "failed to parse {} into token stream: {err}",
          out_rs_path.display()
        )
      });
    assert_display_snapshot!("out.rs", out_rs_contents);
  }

  #[test]
  fn test_foo_handler() {
    let foo_handler_path = Path::new(".openapi-lambda/foo_handler.rs");
    let foo_handler_contents = std::fs::read_to_string(foo_handler_path)
      .unwrap_or_else(|err| panic!("failed to read {}: {err}", foo_handler_path.display()));
    foo_handler_contents
      .parse::<TokenStream>()
      .unwrap_or_else(|err| {
        panic!(
          "failed to parse {} into token stream: {err}",
          foo_handler_path.display()
        )
      });
    assert_display_snapshot!("foo_handler.rs", foo_handler_contents);
  }

  #[test]
  fn test_bar_handler() {
    let bar_handler_path = Path::new(".openapi-lambda/bar_handler.rs");
    let bar_handler_contents = std::fs::read_to_string(bar_handler_path)
      .unwrap_or_else(|err| panic!("failed to read {}: {err}", bar_handler_path.display()));
    bar_handler_contents
      .parse::<TokenStream>()
      .unwrap_or_else(|err| {
        panic!(
          "failed to parse {} into token stream: {err}",
          bar_handler_path.display()
        )
      });
    assert_display_snapshot!("bar_handler.rs", bar_handler_contents);
  }
}
