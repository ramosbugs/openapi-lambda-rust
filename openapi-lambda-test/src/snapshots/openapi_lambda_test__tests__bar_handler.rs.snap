---
source: openapi-lambda-test/src/lib.rs
expression: bar_handler_contents
---
#![allow(unused_imports)]

use crate::bar::{Api, CreateBarResponse};

use openapi_lambda::__private::anyhow;
use openapi_lambda::__private::aws_lambda_events::encodings::Body;
use openapi_lambda::async_trait::async_trait;
use openapi_lambda::{
  ApiGatewayProxyRequestContext, HeaderMap, HttpResponse, LambdaContext, StatusCode,
};

pub struct BarApiHandler {
  // Store any handler state (e.g., DB client) here.
  state: (),
}

impl BarApiHandler {
  pub fn new(state: ()) -> Self {
    Self { state }
  }
}

#[async_trait]
impl Api for BarApiHandler {
  // Define a type here to represent a successfully authenticated user.
  type AuthOk = ();

  // Define an error type to capture the errors produced by your API handler methods.
  type HandlerError = ();

  // Return an error response depending on the nature of the error (e.g., 400 Bad Request for
  // errors caused by a client sending an invalid request, or 500 Internal Server Error for
  // internal errors such as failing to connect to a database).
  async fn respond_to_handler_error(&self, _err: Self::HandlerError) -> HttpResponse {
    todo!()
  }

  async fn create_bar(
    &self,
    bar_id: crate::types::BarId,
    sort_by: Option<crate::models::SortBy>,
    r#type: Option<crate::models::CreateBarTypeParam>,
    x_bar: Option<String>,
    request_body: Vec<u8>,
    headers: HeaderMap,
    request_context: ApiGatewayProxyRequestContext,
    lambda_context: LambdaContext,
    auth_ok: Self::AuthOk,
  ) -> Result<(CreateBarResponse, HeaderMap), Self::HandlerError> {
    todo!()
  }
}

